[PLT-7939] Image thumbnails improvements

saturninoabril opened the pull request:
#### Summary
[Note: This PR is partially completed only. Submitted for further modification on styles.]

Basic functionality for image thumbnails improvement
- added `SingleImageView`
- change layout of multiple files preview
- temporarily fix issue when uploading multiple files (cloning of objects at `FilePreview`)

@asaadmahmood Please help fix CSS styles and add file icons based on UI design

#### Ticket Link
Jira ticket: [PLT-7939](https://mattermost.atlassian.net/browse/PLT-7939)

#### Checklist
[Place an '[x]' (no spaces) in all applicable fields. Please remove unrelated fields.]
- [x] Ran `make check-style` to check for style errors (required for all pull requests)
- [x] Ran `make test` to ensure unit and component tests passed
- [x] Added or updated unit tests (required for all new features) >> test will be updated after fixing CSS styles/file icons, and after approval from PM
- [x] Has UI changes

mattermod commented:
`Setup Test Server` label detected. Spinmint test server created if build succeeds (checks pass and no conflicts with base branch).

mattermod commented:
Spinmint test server created at: http://i-0e5f38a0aec84895b.spinmint.com

Test Account 1: Email: `test@test.com` | Password: `passwd`

Test Account 2: Email: `test2@test.com` | Password: `passwd`

Instance ID: i-0e5f38a0aec84895b

mattermod commented:
Spinmint test server destroyed

mattermod commented:
`Setup Test Server` label detected. Spinmint test server created if build succeeds (checks pass and no conflicts with base branch).

mattermod commented:
Spinmint test server created at: http://i-0d27767a3e0363f35.spinmint.com

Test Account 1: Email: `test@test.com` | Password: `passwd`

Test Account 2: Email: `test2@test.com` | Password: `passwd`

Instance ID: i-0d27767a3e0363f35

asaadmahmood commented:
@esethna @lfbrock Updated the UI with a similar design that Charles had, please have a look. The responsive layout is not fixed, will fix once we are finalised on the desktop one.

mattermod commented:
Spinmint test server destroyed

mattermod commented:
`Setup Test Server` label detected. Spinmint test server created if build succeeds (checks pass and no conflicts with base branch).

mattermod commented:
Spinmint test server created at: http://i-0da7f3adf765d43dc.spinmint.com

Test Account 1: Email: `test@test.com` | Password: `passwd`

Test Account 2: Email: `test2@test.com` | Password: `passwd`

Instance ID: i-0da7f3adf765d43dc

mattermod commented:
Spinmint test server destroyed

mattermod commented:
`Setup Test Server` label detected. Spinmint test server created if build succeeds (checks pass and no conflicts with base branch).

mattermod commented:
Spinmint test server created at: http://i-09dec063c838cfc4b.spinmint.com

Test Account 1: Email: `test@test.com` | Password: `passwd`

Test Account 2: Email: `test2@test.com` | Password: `passwd`

Instance ID: i-09dec063c838cfc4b

mattermod commented:
Spinmint test server destroyed

mattermod commented:
`Setup Test Server` label detected. Spinmint test server created if build succeeds (checks pass and no conflicts with base branch).

mattermod commented:
Spinmint test server created at: http://i-00e8d8e5d56ac98cf.spinmint.com

Test Account 1: Email: `test@test.com` | Password: `passwd`

Test Account 2: Email: `test2@test.com` | Password: `passwd`

Instance ID: i-00e8d8e5d56ac98cf

mattermod commented:
Spinmint test server destroyed

mattermod commented:
`Setup Test Server` label detected. Spinmint test server created if build succeeds (checks pass and no conflicts with base branch).

asaadmahmood commented:
Done, apart from 3,4. Those need to be addressed by @saturninoabril

mattermod commented:
Spinmint test server destroyed

mattermod commented:
`Setup Test Server` label detected. Spinmint test server created if build succeeds (checks pass and no conflicts with base branch).

mattermod commented:
Spinmint test server created at: http://i-095676c57a77a80df.spinmint.com

Test Account 1: Email: `test@test.com` | Password: `passwd`

Test Account 2: Email: `test2@test.com` | Password: `passwd`

Instance ID: i-095676c57a77a80df

mattermod commented:
`Setup Test Server` label detected. Spinmint test server created if build succeeds (checks pass and no conflicts with base branch).

mattermod commented:
Spinmint test server created at: http://i-0fbedbf2cb8c6f0c0.spinmint.com

Test Account 1: Email: `test@test.com` | Password: `passwd`

Test Account 2: Email: `test2@test.com` | Password: `passwd`

Instance ID: i-0fbedbf2cb8c6f0c0

saturninoabril commented:
@esethna 
- no. 3, I think this is the current behavior for image download
- no. 4, done. 
- no. 5, done. @asaadmahmood please help fix SVG image preview size if that's too small.

mattermod commented:
Spinmint test server destroyed

mattermod commented:
`Setup Test Server` label detected. Spinmint test server created if build succeeds (checks pass and no conflicts with base branch).

mattermod commented:
Spinmint test server created at: http://i-0b16890fb4c1d88b8.spinmint.com

Test Account 1: Email: `test@test.com` | Password: `passwd`

Test Account 2: Email: `test2@test.com` | Password: `passwd`

Instance ID: i-0b16890fb4c1d88b8

esethna commented:
Thanks @saturninoabril @asaadmahmood. Another review pass after the latest changes:

1. Scroll pop and scroll location issues. Switch channels to one with single image thumbnails. Notice scroll pop everytime the channel comes into view. Also notice that the channel does not load at the bottom, it requires manual scrolling to reach the bottom of the channel. This only reproduces in channels with single image thumbnails, not multiple thumbnail uploads.

![scrollpop](https://user-images.githubusercontent.com/14333569/33496081-19eff588-d686-11e7-985f-f3e3dd917c14.gif)

2. No hover effect on the download button for multiple image thumbnails

![hover](https://user-images.githubusercontent.com/14333569/33496200-8c4765ee-d686-11e7-8e22-d88851ab16a4.gif)

3. Title text for single image thumbnails still seems large and obtrusive to conversation flow: 
![image](https://user-images.githubusercontent.com/14333569/33496249-ba9534e4-d686-11e7-8bc8-28f86549784d.png)

asaadmahmood commented:
@esethna Once Charles come up with a redesign for the single image uploads to solve the title issue, let me know. I'll push the hover fix along with that, and I'll ask Saturn to look at the scroll pop issue.

saturninoabril commented:
@asaadmahmood Fix loading of single view image. Also added `loading` class which is basically empty just in case you'd like to modify loading styles.

mattermod commented:
`Setup Test Server` label detected. Spinmint test server created if build succeeds (checks pass and no conflicts with base branch).

mattermod commented:
Spinmint test server created at: http://i-03d901d4511894c74.spinmint.com

Test Account 1: Email: `test@test.com` | Password: `passwd`

Test Account 2: Email: `test2@test.com` | Password: `passwd`

Instance ID: i-03d901d4511894c74

esethna commented:
FYI, reached out to AM and CH on a few more updates

jasonblais commented:
@asaadmahmood Is this ready for re-testing?

saturninoabril commented:
@jasonblais I believe this is ready for PM review. Please give me a minute to rebase and fix merge conflicts.

mattermod commented:
Spinmint test server destroyed

mattermod commented:
`Setup Test Server` label detected. Spinmint test server created if build succeeds (checks pass and no conflicts with base branch).

mattermod commented:
Spinmint test server created at: http://i-0e7f04537eaf0c921.spinmint.com

Test Account 1: Email: `test@test.com` | Password: `passwd`

Test Account 2: Email: `test2@test.com` | Password: `passwd`

Instance ID: i-0e7f04537eaf0c921

jasonblais commented:
Thanks Saturn. I'll let Eric test it next week when he's back, since he's most familiar with the PR.

lfbrock commented:
@esethna - it seems like there's a lot of un-clickable space on the non-image file thumbnails that doesn't feel natural/"forgiving". Can you work with Mike on the UX there to get his thoughts?

kylegoch commented:
Was brought here from https://github.com/mattermost/mattermost-server/issues/8087

Used the test server and this was the action I was hoping for more in my issue.

I have a few suggestions:
1. When uploading multiple pictures they should render the same as if they were a single image. So instead of going into a line of boxes, it should instead just break out and display each image.
2. When files (or multiple images) are uploaded the box has a lot of whitespace and look kind of weird.
3. The image header is rather large, and should be clickable to open the image.

Just my two cents.

esethna commented:
Thanks for the feedback @kylegoch!

@lfbrock ya I agree, I think overall there are some inconsistencies in the hover effects and click boxes. I have a meeting with MG and CH tomorrow, we can chat about this to get to a plan.

esethna commented:
@saturninoabril chatted with Charles (CH) and MG, we want to make the following adjustments:

1. Make the entire thumbnail card clickable to open the preview window (CH is working on a new hover effect for this), and make only the download button trigger a download. 

![image](https://user-images.githubusercontent.com/14333569/35068067-31146190-fb93-11e7-8fb2-e5b70436cf4f.png)

2. The title for single image uploads should also be clickable to open the preview window (when hovering the text should underline)

![image](https://user-images.githubusercontent.com/14333569/35068129-60180bc2-fb93-11e7-90f9-6e338a9182a7.png)

3. The download button on single image thumbnails is missing a hover effect (CH is working on updating the design atm)

![image](https://user-images.githubusercontent.com/14333569/35068273-d3b45216-fb93-11e7-979b-d63ec3fbb2d1.png)

saturninoabril commented:
@asaadmahmood I did rebase to latest master and fixed merge conflicts. I've addressed comment 1 and 2 but I'm getting to where on CSS styles. Please help fix it. Thanks!

mattermod commented:
Spinmint test server destroyed

mattermod commented:
`Setup Test Server` label detected. Spinmint test server created if build succeeds (checks pass and no conflicts with base branch).

mattermod commented:
Spinmint test server created at: http://i-00d117c9af3717ed8.spinmint.com

Test Account 1: Email: `test@test.com` | Password: `passwd`

Test Account 2: Email: `test2@test.com` | Password: `passwd`

Instance ID: i-00d117c9af3717ed8

asaadmahmood commented:
@esethna @saturninoabril Done.

mattermod commented:
Spinmint test server destroyed

mattermod commented:
`Setup Test Server` label detected. Spinmint test server created if build succeeds (checks pass and no conflicts with base branch).

mattermod commented:
Spinmint test server created at: http://i-06c60616d4ae2c40b.spinmint.com

Test Account 1: Email: `test@test.com` | Password: `passwd`

Test Account 2: Email: `test2@test.com` | Password: `passwd`

Instance ID: i-06c60616d4ae2c40b

mattermod commented:
Spinmint test server destroyed

mattermod commented:
`Setup Test Server` label detected. Spinmint test server created if build succeeds (checks pass and no conflicts with base branch).

mattermod commented:
Spinmint test server created at: http://i-02a5ac9079aff05d7.spinmint.com

Test Account 1: Email: `test@test.com` | Password: `passwd`

Test Account 2: Email: `test2@test.com` | Password: `passwd`

Instance ID: i-02a5ac9079aff05d7

mattermod commented:
Spinmint test server destroyed

stephenkiers commented on components/svg/download_icon.jsx:
```diff
@@ -0,0 +1,41 @@
+    render() {
+        return (
+            <span {...this.props}>
+                <svg
```
why is this a component, when the rest are svgs?

hmhealey replied:
With how the other SVGs are imported, I think it's fine to keep them as SVG files instead of components since that'll likely make it easier if they need to be changed in the future. I forgot that was even an option at the time

stephenkiers commented on images/icons/audio.svg:
```diff
@@ -0,0 +1,24 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<svg width="24px" height="32px" viewBox="0 0 24 32" version="1.1" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink">
+    <!-- Generator: Sketch 47 (45396) - http://www.bohemiancoding.com/sketch -->
```
all the garbage metadata could probably be stripped from these svgs. Perhaps that is a separate issue though, as someone with SVG knowledge should do that to all the svgs.

saturninoabril commented:
hmmm, wait... there's something wrong with my rebase...

saturninoabril commented:
Rebase issue is now solved after it went on a lost path to master. Thanks to @jespino for helping out!

@stephenkiers This is now ready for dev review.
